typing/zero_alloc.ml

```diff
@@ -161,23 +163,35 @@ let sub_const_const_exn za1 za2 =
   | None, None -> ()
 
 let sub_var_const_exn v c =
-  (* This can only fail due to an arity mismatch. We have a linear order and can
+  (* This can only fail due to an arity  mismatch. We have a linear order and can
      always constrain the var lower to make the sub succeed. *)
   match v, c with
   | _, (Default_zero_alloc | Ignore_assert_all | Assume _) -> assert false
   | { arity = arity1; _ }, Check { arity = arity2; _ }
     when arity1 <> arity2 ->
     raise (Error (Arity_mismatch (arity1, arity2)))
-  | { desc = None; _ }, Check { strict; opt; _ } ->
+  | { desc = None; _ }, Check { strict; opt; custom_error_msg;  } ->
     !log_change (None, v);
-    v.desc <- Some { strict; opt }
-  | { desc = (Some { strict = strict1; opt = opt1 } as desc); _ },
-    Check { strict = strict2; opt = opt2 } ->
+    v.desc <- Some { strict; opt; custom_error_msg }
+  | { desc = (Some { strict = strict1; opt = opt1; custom_error_msg = msg1; } as desc); _ },
+    Check { strict = strict2; opt = opt2; custom_error_msg = msg2 } ->
     let strict = strict1 || strict2 in
     let opt = opt1 && opt2 in
-    if strict <> strict1 || opt <> opt1 then begin
+    let custom_error_msg, msg_changed =
+      match msg1, msg2 with
+      | None, None -> msg1, false;
+      | None, Some _ -> msg2, true;
+      | Some _, None -> msg1, false;
+      | Some m1, Some m2 ->
+        let b = String.equal m1 m2 in
+        let msg =
+          if b then msg1 else Some (String.concat "\n" [m1; m2])
+        in
+        msg, b
```

Comment:
I think `b` here is flipped - it's true if we're not changing the message, but should be true if we are.  Here is a program you can add to the testsuite that should demonstrate the concatenated error message once this is fixed:
```ocaml
module type S1 = sig
  val f : 'a -> ('a * 'a) [@@zero_alloc custom_error_message "foo"]
end

module type S2 = sig
  val f : 'a -> ('a * 'a) [@@zero_alloc custom_error_message "bar"]
end

module M = struct
  let f x = (x,x)
end

module M' : S1 = M

module M'' : S2 = M
```

Reply:
I guess one other point to make about this:  Because this concatenation behavior only happens in the case where the thing on the left is a variable, we'll often end up throwing away the string.  For example, if the custom error message appears in a signature but there is also a zero_alloc annotation in the structure (with or without a custom error message), we'll just throw away the string.

I think we could fix this by making the `custom_error_message` field itself a ref and just mutating that.  But you don't necessarily have to do that in this first version.

Reply:
> I think b here is flipped - it's true if we're not changing the message, but should be true if we are.
thanks!! fixed.

I added the test you suggested to `typing-zero-alloc/signatures` but the `%expect` output it's not showing the custom error on `f`. I know the custom error gets propagated correctly the backend because when the new test is compiled, the zero alloc check fails with the combined string in the output.